src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/mappers/FlowPathMapper.java

```diff
@@ -65,28 +68,27 @@ public PathInfoData map(FlowPath path) {
         List<PathNodePayload> resultList = new ArrayList<>();
 
         Flow flow = flowPath.getFlow();
-        int srcPort = flow.isForward(flowPath) ? flow.getSrcPort() : flow.getDestPort();
-        int destPort = flow.isForward(flowPath) ? flow.getDestPort() : flow.getSrcPort();
-
-        if (flowPath.getSegments().isEmpty()) {
-            resultList.add(
-                    new PathNodePayload(flowPath.getSrcSwitch().getSwitchId(), srcPort, destPort));
+        FlowEndpoint ingress = FlowSideAdapter.makeIngressAdapter(flow, flowPath).getEndpoint();
+        FlowEndpoint egress = FlowSideAdapter.makeEgressAdapter(flow, flowPath).getEndpoint();
+
+        List<PathSegment> pathSegments = flowPath.getSegments();
+        Iterator<PathSegment> leftIter = pathSegments.iterator();
+        Iterator<PathSegment> rightIter = pathSegments.iterator();
+        if (! rightIter.hasNext()) {
+            resultList.add(new PathNodePayload(
+                    flowPath.getSrcSwitch().getSwitchId(), ingress.getPortNumber(), egress.getPortNumber()));
         } else {
-            List<PathSegment> pathSegments = flowPath.getSegments();
-
-            resultList.add(new PathNodePayload(flowPath.getSrcSwitch().getSwitchId(), srcPort,
-                    pathSegments.get(0).getSrcPort()));
-
-            for (int i = 1; i < pathSegments.size(); i++) {
-                PathSegment inputNode = pathSegments.get(i - 1);
-                PathSegment outputNode = pathSegments.get(i);
-
-                resultList.add(new PathNodePayload(inputNode.getDestSwitch().getSwitchId(), inputNode.getDestPort(),
-                        outputNode.getSrcPort()));
+            PathSegment left;
```

Comment:
2 questions: 
- How are the changes in this file related to the title "NPE during producting flow dump"?
- What benefits do these changes bring? IMHO, 2 iterators instead of an index are neither shorter nor clearer.

Reply:
The only reason I touch this code is to remove call flow.isForward(flowPath) because this call produces an incorrect answer for "extra/detached" paths (we inject such path now into RevertResourceAllocationAction). All other changes related to this one... I have tried to make it simpler, looks like I have failed this target.

If you insist I will revert iteration by indexes.

Reply:
Got it.